public/app/features/dashboard/state/runRequest.ts

```diff
@@ -0,0 +1,172 @@
+// Libraries
+import { Observable, of, timer, merge } from 'rxjs';
+import { flatten, map as lodashMap, isArray, isString } from 'lodash';
+import { map, catchError, takeUntil, mapTo, share, finalize } from 'rxjs/operators';
+
+// Utils & Services
+import { LoadingState, dateMath } from '@grafana/data';
+import { getBackendSrv } from 'app/core/services/backend_srv';
+
+// Types
+import { DataSourceApi, DataQueryRequest, PanelData, DataQueryResponsePacket, DataQueryError } from '@grafana/ui';
+
+type MapOfResponsePackets = { [str: string]: DataQueryResponsePacket };
+
+interface RunningQueryState {
+  packets: { [key: string]: DataQueryResponsePacket };
+  panelData: PanelData;
+}
+
+/*
+ * This function should handle composing a PanelData from multiple responses
+ * Does not handle deltas yet
```

Comment:
With Loki change to send full frames on streaming (and no other DS using delta frames as per @ryantxu comment) is there need to handle them? Can we just tell datasource devs to return full frames as part of the API contract?

Reply:
yea, can remove that comment